feat: add pull flow testing (#504)
FilOzone/dealbot: docs/checks/pull-check.md

# Pull Check

This document is the **source of truth** for how dealbot's Pull check works.

Source code links throughout this document point to the current implementation.

For event and metric definitions used by the dashboard, see [Dealbot Events & Metrics](./events-and-metrics.md).

## Overview

A "pull check" exercises the **storage provider pull-to-park pathway**: dealbot publishes a temporary piece at `/api/piece/{pieceCid}`, asks the SP to fetch (pull) and park it via the Synapse `pullPieces` API, waits for a terminal SP pull status, and finally re-fetches the piece from the SP to verify byte-for-byte integrity.

> The pull check answers a different question than the [Data Storage check](./data-storage.md): instead of *uploading* bytes to the SP, it asks the SP to *pull* bytes from a public URL. This validates an SP's outbound HTTP fetcher, the pull request lifecycle, and retrieval surface.

Reviewer: ```suggestion
The pull check answers a different question than the [Data Storage check](./data-storage.md): instead of *uploading* bytes to the SP, it asks the SP to *pull* bytes from a public URL. This validates an SP's outbound HTTP fetcher, the pull request lifecycle, and `/piece` retrieval surface. (`/piece` retrieval is not covered by the [Data Storage check](./data-storage.md).)
```

A successful pull check requires all [assertions in the table below](#what-gets-asserted) to pass. Failure occurs if any step fails or the job exceeds its max allowed time. Operational timeouts exist to prevent jobs from running indefinitely, but they are not quality assertions.

> > **Where results live:** Pull check results are exported to Prometheus and structured logs only. They are **not** persisted in Postgres or written to ClickHouse.

Reviewer: Ideally there wouldn't be any special callouts here. Any issue with us writing to clickhouse?

Author: https://github.com/FilOzone/dealbot/pull/522

## What Gets Asserted

Each pull check asserts the following for every SP:

| # | Assertion | How It's Checked | Retries | Relevant Metric for Setting a Max Duration | Implemented? |
|---|-----------|------------------|:---:|--------------------------------------------|:---:|

> | 1 | SP accepts the pull request | `pullPieces` returns without error and reports a non-terminal-failure status | 0 | [`pullCheckRequestLatencyMs`](./events-and-metrics.md#pullCheckRequestLatencyMs) | Yes |

Reviewer: ```suggestion
| 1 | SP accepts the pull request | Synapse `pullPieces` (i.e., initial call to SP `POST /pdp/piece/pull`) returns without error and reports a non-terminal-failure status | 0 | [`pullCheckRequestLatencyMs`](./events-and-metrics.md#pullCheckRequestLatencyMs) | Yes |
```

> | 2 | SP reaches a terminal `complete` pull status | `waitForPullStatus` polls the SP until a terminal status is reported | Polling with delay until [`PULL_CHECK_JOB_TIMEOUT_SECONDS`](../environment-variables.md#pull_check_job_timeout_seconds) | [`pullCheckCompletionLatencyMs`](./events-and-metrics.md#pullCheckCompletionLatencyMs) | Yes |

Reviewer: ```suggestion
| 2 | SP reaches a terminal `complete` pull status | Synapse `waitForPullStatus` polls the SP (using `POST /pdp/piece/pull`) until a terminal status is reported | Polling with contune until [`PULL_CHECK_JOB_TIMEOUT_SECONDS`](../environment-variables.md#pull_check_job_timeout_seconds) | [`pullCheckCompletionLatencyMs`](./events-and-metrics.md#pullCheckCompletionLatencyMs) | Yes |
```

| 3 | SP serves the pulled piece via `/piece/{pieceCid}` | Re-fetch the bytes from the SP's PDP service URL and re-compute the piece CID | 0 | n/a (bounded by job timeout) | Yes |
| 4 | All checks pass | Pull check is not marked successful until all assertions pass within the job timeout | n/a | [`pullCheckCompletionLatencyMs`](./events-and-metrics.md#pullCheckCompletionLatencyMs) | Yes |

## Pull Check Lifecycle

The dealbot scheduler triggers pull check jobs at a configurable rate (`PULL_CHECKS_PER_SP_PER_HOUR`).

> 
> ```mermaid
> flowchart TD
>   Generate["Compute PieceCID + register hosted source in Postgres<br/>at /api/piece/{pieceCid}"]
>   Generate --> Submit["Submit pullPieces request to SP"]
>   Submit --> Poll["Poll SP via waitForPullStatus<br/>until terminal pull status"]
>   Poll -->|complete| Validate["Direct /piece/{pieceCid} fetch from SP<br/>+ recompute pieceCid"]
>   Poll -->|other terminal status| Fail["Mark pull check failed"]
>   Validate -->|matches| Success["Mark pull check successful"]
>   Validate -->|mismatch or fetch error| Fail
>   Success --> Cleanup
>   Fail --> Cleanup["Forget hosted piece registration"]
> ```

Reviewer: ```suggestion
```mermaid
flowchart TD
  Generate["Compute PieceCID + register hosted source in Postgres<br/>at /api/piece/{pieceCid}"]
  Generate --> Submit["Submit pullPieces request to SP"]
  Submit --> |SP responds with HTTP 200|Poll["Poll SP via waitForPullStatus<br/>until terminal pull status"]
  Submit --> |SP doesn't respond with HTTP 200| Fail["Mark pull check failed"]
  Poll -->|complete| Validate["Direct /piece/{pieceCid} fetch from SP<br/>+ recompute pieceCid"]
  Poll -->|other terminal status| Fail
  Validate -->|matches| Success["Mark pull check successful"]
  Validate -->|mismatch or fetch error| Fail
  Success --> Cleanup
  Fail --> Cleanup["Forget hosted piece registration"]
```
```

My understanding is that there is another failure that can occur if the initial pull request isn't accepted.

### 1. Prepare the hosted piece

> Dealbot computes a deterministic PieceCID for a synthetic test piece and registers it in the Postgres `pull_pieces` table. The registration carries a TTL controlled by `PULL_CHECK_HOSTED_PIECE_TTL_SECONDS` so the source remains available for the entire pull window.

Reviewer 2: `PULL_CHECK_HOSTED_PIECE_TTL_SECONDS` var is not implemented? only mention is in this doc in two places.

By persisting registrations to Postgres instead of in-memory, the hosted source can be resolved by any API pod in a multi-pod deployment, even if the pull check was initiated by a different worker pod.

> The synthetic data is **not** stored on disk. Instead, dealbot uses a deterministic pseudo-random generator (AES-256-CTR) to stream the same bytes whenever the SP fetches the piece or dealbot needs to re-compute the CID for validation.

Reviewer: ```suggestion
Dealbot computes a deterministic PieceCID for a synthetic test piece.  The synthetic data is **not** stored on disk. Instead, dealbot uses a deterministic pseudo-random generator (AES-256-CTR) to generate it as needed. 

The pieceCid and synthetic piece's random seed are registered it in the Postgres `pull_pieces` table. The registration carries a TTL controlled by `PULL_CHECK_HOSTED_PIECE_TTL_SECONDS` so the source remains available for the entire pull window.

By persisting registrations to Postgres instead of in-memory, the hosted source can be resolved by any API pod in a multi-pod deployment, even if the pull check was initiated by a different worker pod. It generates the same piece and streams the whenever the SP fetches the piece or dealbot needs to re-compute the CID for validation.
```

Reviewer: I think we should we should talk about how the pieceCid is constructed before we talk about storing it to the DB.

The source URL handed to the SP is built from the dealbot `app.apiPublicUrl` config (set via `DEALBOT_API_PUBLIC_URL`). When `DEALBOT_API_PUBLIC_URL` is unset, dealbot falls back to `http://{DEALBOT_HOST}:{DEALBOT_PORT}`, which is only reachable in single-host or `localhost` setups.

- **Data format:** Deterministic pseudo-random bytes
- **Default size:** `PULL_CHECK_PIECE_SIZE_BYTES` (default 10 MiB)
- **Source URL:** `{apiPublicUrl}/api/piece/{pieceCid}`

Source: [`pull-check.service.ts` (`preparePullPiece`)](../../apps/backend/src/pull-check/pull-check.service.ts), [`pull-piece.repository.ts`](../../apps/backend/src/pull-check/pull-piece.repository.ts)

### 2. Submit the pull request

> Dealbot calls `pullPieces` from `@filoz/synapse-core/sp` with the pieceCid, the source URL, and either the SP's existing `dataSetId`/`clientDataSetId` or the SP `payee` for new-dataset flows. The submission timestamp is stamped on the registration so it can later be subtracted from the first-byte event.

Reviewer: what is " the SP's existing `dataSetId`/`clientDataSetId` or the SP `payee` for new-dataset flows"? This is the first I am seeing this discussed in this doc. This needs expansion.

Author: This section is a bit outdated. We no longer pass `dataSetId`/`clientDataSetId`. Instead, we pass the `payee`, which is required by the sdk for validation inside curio. curio then authenticates it via an `eth_call` to the pdp verifier contract.

Reviewer: Ok, that's a good detail to maybe add to this section? Basically when we make the `pullPieces` call, we pass Dealbot as the `payee` to pass Curio's validation... ?

Source: [`pull-check.service.ts` (`runPullCheck`)](../../apps/backend/src/pull-check/pull-check.service.ts)

### 3. Wait for terminal SP pull status

`waitForPullStatus` polls the SP at `PULL_CHECK_POLL_INTERVAL_SECONDS` until the SP reports a terminal status (`complete` or `failed`) or the job timeout fires. Dealbot increments the [`pullCheckProviderStatus`](./events-and-metrics.md#pullCheckProviderStatus) counter exactly once with the **terminal** status; intermediate poll statuses are not counted.

> When the SP fetches `/api/piece/{pieceCid}` for the first time, the controller stamps a first-byte timestamp on the registration. This is the basis for [`pullCheckFirstByteMs`](./events-and-metrics.md#pullCheckFirstByteMs).

Reviewer: ```suggestion
When dealbot receives the SP request for `/api/piece/{pieceCid}` for the first time, dealbot stamps a first-byte timestamp on the registration. This is the basis for [`pullCheckFirstByteMs`](./events-and-metrics.md#pullCheckFirstByteMs).

In parallel, dealbot`waitForPullStatus` polls the SP at `PULL_CHECK_POLL_INTERVAL_SECONDS` until the SP reports a terminal status (`complete` or `failed`) or the job timeout fires. Dealbot increments the [`pullCheckProviderStatus`](./events-and-metrics.md#pullCheckProviderStatus) counter exactly once with the **terminal** status; intermediate poll statuses are not counted.

```

I removed the "controller" logic, unless that is covered elsewhere.

I rearranged the paragraphs to go in chronological order of the events.

Source: [`pull-piece.controller.ts`](../../apps/backend/src/pull-check/pull-piece.controller.ts)

### 4. Direct piece-fetch validation

After commit, dealbot fetches `{serviceURL}/piece/{pieceCid}` from the SP, re-computes the piece CID over the response body, and compares it against the expected CID. A mismatch fails the pull check with `failure.other`. A network or HTTP error during validation also fails the check (transport errors are intentionally not retried).

> Aborts (job timeout) propagate as throws and are classified as `failure.timedout` rather than as a validation failure.

Reviewer: ```suggestion
After `waitForPullStatus`, dealbot fetches `{serviceURL}/piece/{pieceCid}` from the SP, re-computes the piece CID over the response body, and compares it against the expected CID. A mismatch fails the pull check with `failure.other`. A network or HTTP error during validation also fails the check (transport errors are intentionally not retried).

Aborts (job timeout) propagate as throws and are classified as `failure.timedout` rather than as a validation failure.
```

I wasn't sure what was meant by "commit".

Source: [`pull-check.service.ts` (`validateByDirectPieceFetch`)](../../apps/backend/src/pull-check/pull-check.service.ts)

### 5. Cleanup

> Whether the pull check succeeds or fails, the `finally` block removes the registration entry.

Reviewer: ```suggestion
Whether the pull check succeeds or fails, the `finally` block removes the registration entry from postgres' database.
```

After cleanup, subsequent `/api/piece/{pieceCid}` requests return HTTP 404 Not Found.

Source: [`pull-check.service.ts`](../../apps/backend/src/pull-check/pull-check.service.ts)

## Pull Check Status Progression

A pull check has a single terminal status, recorded once per check via [`pullCheckStatus`](./events-and-metrics.md#pullCheckStatus):

| Overall Status | Meaning |
|--------|---------|

> | `success` | All five [assertions](#what-gets-asserted) passed within the job timeout. |

Reviewer 2: there are only 4 assertions. remove count 
```suggestion
| `success` | All [assertions](#what-gets-asserted) passed within the job timeout. |
```

| `failure.timedout` | The job was aborted because it exceeded `PULL_CHECK_JOB_TIMEOUT_SECONDS`, or the underlying error message indicates a timeout. |
| `failure.other` | Any other failure: SP rejected the pull request, SP reached a non-`complete` terminal status, or direct piece validation failed. |

Failures are classified by inspecting the error message; see [`classifyFailureStatus`](../../apps/backend/src/metrics-prometheus/check-metric-labels.ts) for the exact rule.

In addition to the overall status, dealbot records the **raw SP-reported terminal pull status** via [`pullCheckProviderStatus`](./events-and-metrics.md#pullCheckProviderStatus) (for example `complete`, `failed`, `not_found`). This separates "SP said it failed" from "dealbot's downstream validation failed" in dashboards.

## HTTP API

The dealbot API exposes one endpoint dedicated to pull checks:

| Method | Path | Description |
|--------|------|-------------|

> | `GET` | `/api/piece/{pieceCid}` | Streams the temporary hosted piece bytes for an in-flight pull check. Returns `200` with the bytes when an active registration exists, `410 Gone` when the registration has been cleaned up or expired, and `404 Not Found` when no registration exists. |

Reviewer 2: Code only has 404 currently?

> The endpoint is registered on the same `/api` prefix as the other dealbot HTTP endpoints. It is intentionally unauthenticated because SPs must be able to pull from it during a check; access is bounded by the per-piece TTL.

Reviewer 2: missing TTL env var

Source: [`pull-piece.controller.ts`](../../apps/backend/src/pull-check/pull-piece.controller.ts)

## Metrics Recorded

Metric definitions (including Prometheus metrics) live in [Dealbot Events & Metrics](./events-and-metrics.md). The metrics emitted by a pull check are:

- [`pullCheckRequestLatencyMs`](./events-and-metrics.md#pullCheckRequestLatencyMs)
- [`pullCheckCompletionLatencyMs`](./events-and-metrics.md#pullCheckCompletionLatencyMs)
- [`pullCheckFirstByteMs`](./events-and-metrics.md#pullCheckFirstByteMs) (only when the SP actually pulled from `/api/piece/{pieceCid}`)
- [`pullCheckThroughputBps`](./events-and-metrics.md#pullCheckThroughputBps)
- [`pullCheckStatus`](./events-and-metrics.md#pullCheckStatus)

> - [`pullCheckProviderStatus`](./events-and-metrics.md#pullCheckProviderStatus)

Reviewer: ```suggestion
- [`pullCheckRequestLatencyMs`](./events-and-metrics.md#pullCheckRequestLatencyMs)
- [`pullCheckFirstByteMs`](./events-and-metrics.md#pullCheckFirstByteMs) (only when the SP actually pulled from `/api/piece/{pieceCid}`)
- [`pullCheckCompletionLatencyMs`](./events-and-metrics.md#pullCheckCompletionLatencyMs)
- [`pullCheckThroughputBps`](./events-and-metrics.md#pullCheckThroughputBps)
- [`pullCheckStatus`](./events-and-metrics.md#pullCheckStatus)
- [`pullCheckProviderStatus`](./events-and-metrics.md#pullCheckProviderStatus)
```

I like to discuss things in chronologoical order.

And per below, what is the scenario where an SP won't call `/api/piece/{pieceCid}`?

Author: If the initial pull piece POST request fails due to sp fault, then SP won't call `/api/piece/{pieceCid}`.

## Configuration

Key environment variables that control pull check behavior:

| Variable | Description |
|----------|-------------|
| `DEALBOT_API_PUBLIC_URL` | Public base URL used to construct the hosted-piece source URL handed to SPs. Required for any deployment where SPs cannot reach `DEALBOT_HOST:DEALBOT_PORT` directly. |
| `PULL_CHECKS_PER_SP_PER_HOUR` | Per-SP pull check rate. |
| `PULL_CHECK_JOB_TIMEOUT_SECONDS` | Max end-to-end pull check job runtime before forced abort. |

> | `PULL_CHECK_HOSTED_PIECE_TTL_SECONDS` | TTL of the temporary hosted piece source served at `/api/piece/{pieceCid}`. |

Reviewer 2: non-existent in code?

| `PULL_CHECK_POLL_INTERVAL_SECONDS` | Polling interval used while waiting for a terminal SP pull status. |
| `PULL_CHECK_PIECE_SIZE_BYTES` | Size of the synthetic test piece dealbot generates per pull check. |

Source: [`apps/backend/src/config/app.config.ts`](../../apps/backend/src/config/app.config.ts)

See also: [`docs/environment-variables.md`](../environment-variables.md) for the source-of-truth configuration reference.

## FAQ

### Why isn't the pull-check piece tracked in the `deals` table?

Pull checks are intentionally isolated from the data-storage flow: they don't pass through `DealService.createDeal`, don't allocate a `Deal` entity. This keeps the pull-check signal independent of the data-storage success rate.

### Why does a "cached pull" not record `pullCheckFirstByteMs`?

> If an SP previously pulled the same piece CID and serves the new pull request from a local cache, it will never fetch `/api/piece/{pieceCid}`, so dealbot has no first-byte timestamp to subtract. In that case dealbot skips the histogram observation rather than emit a misleading zero. Cached pulls are uncommon today because each pull check generates a fresh random piece, but the registry's first-byte capture is **idempotent** so retried pulls during a single check do not skew measurements either.

Reviewer: What would even cause this to happen? Don't we generate a new CID for each pull check? If this an edge case because we happen to generate the same seed, then I think we can skip this here and in code. That's an extreme edge case (assuming our seed space is large).

Author: Yes, this is extreme edge case will remove it from here and in code.

# Pull Check

This document is the **source of truth** for how dealbot's Pull check works.

Source code links throughout this document point to the current implementation.

For event and metric definitions used by the dashboard, see [Dealbot Events & Metrics](./events-and-metrics.md).

## Overview

> A "pull check" exercises the **storage provider pull-to-park pathway**: dealbot publishes a temporary piece at `/api/piece/{pieceCid}`, asks the SP to fetch (pull) and park it via the Synapse `pullPieces` API, waits for a terminal SP pull status, and finally re-fetches the piece from the SP to verify byte-for-byte integrity.

Reviewer: ```suggestion
A "pull check" exercises the **[storage provider pull-to-park pathway](https://docs.filecoin.cloud/developer-guides/storage/upload-pipeline/#pull-phase-sp-to-sp-transfer)**: dealbot publishes a temporary piece at `/api/piece/{pieceCid}`, asks the SP to fetch (pull) and park it via the Synapse `pullPieces` API, waits for a terminal SP pull status, and finally re-fetches the piece from the SP to verify byte-for-byte integrity.
```
